# For the first part, I am providing two separate solutions:

## 1) the first solution is to scape the saved html webpage as a file using my own method
## 2) the second solution is to use Pandas

In [1]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

import geocoder

from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

import folium # map rendering library

print('Libraries imported.')

Libraries imported.


This first part uses a generator (generating function) to return the lines in the html file based on the start line and stop line

The second function strips off the html text

In [2]:
def extract_data_lines(filename,start_text,end_text):
    """
    open 'filename', and yield the lines between
    the line that contains, 'start_text' and the
    line that contains 'end_text'
    """
    started = False
    ended = False
    with open(filename,encoding="utf8") as fh: #this means that after it is done reading the file it will close it automatically.
        for line in fh:
            #print(line)
            if start_text in line:
                started=True
                #print(line)
                yield line
            elif (started):
                if end_text in line:
                    ended=True
                    yield line
                    break #the break gets out of the loop over lines in the file
                elif (not ended):
                    yield line
                    
def rm_html_jnk(var):
    if ('title=' in var):
        ii = var.find('">')
        ij = var.find('</a>')
        var = var[ii+2:ij]
        return var
    else:
        if ('<td>' in var) and ('</td>' in var):
            ii = var.find('<td>')
            ij = var.find('</td>')
            var = var[ii+4:ij]
            return var
        if ('<td>' in var) and ('</td>' not in var):
            ii = var.find('<td>')
            var = var[ii+4:]
            return var.split('\n')[0]

In [3]:
filename = 'List_of_postal_codes_of_Canada_M.html'
start_text = '<td>M1A</td>'
end_text = '</tbody><tfoot></tfoot></table>'

data = {'Postcode': [],'Borough': [], 'Neighborhood': []}
lst = ['Postcode','Borough','Neighborhood']
post = {}
tr = False

counter = 0 #since the html file splits the row into three separate parts for each column

idx = 0     #since there are repeat postalcodes this index variable keeps track of unique one in a dictionary

temp = []   #this list gets reset every time it hits a new row of variables

same = False #variable to say whether or not new postalcode is already in the dictionary

for line in extract_data_lines(filename,start_text,end_text):
    if end_text not in line:
        #print(temp)
        if '<tr>' in line or '</tr>' in line: #these tags mark the beginning/end of the column variables for the row
            counter = 0
            temp = []
            same = False
        else:
            trim_line = rm_html_jnk(line)
            #temp.append(trim_line)
            if(counter == 0):
                temp.append(trim_line)
                if trim_line in post:
                    same = True
                else:
                    post[trim_line]=idx
                    data[lst[counter]].append(trim_line)
                    idx+=1
            elif(counter == 1):
                temp.append(trim_line)
                if not same:
                    data[lst[counter]].append(rm_html_jnk(line))
            elif(counter == 2):
                temp.append(trim_line) 
                #Below I reassign the Neighbourhood variable if the Borough variable is set
                if trim_line =='Not assigned' and temp[1] != 'Not assigned':
                    trim_line = temp[1]
                    print('resetting neighborhood from {} to {}'.format(temp[2],trim_line))
                if same: 
                    #We can just add the neighborhoods to all the others of the same postalcode
                    data[lst[counter]][post[temp[0]]] = data[lst[counter]][post[temp[0]]]+', '+trim_line
                else:
                    data[lst[counter]].append(trim_line)
            counter+=1

df = pd.DataFrame(data)

resetting neighborhood from Not assigned to Queen's Park


### As you can see below, this was properly read in except for the fact that we need to drop the rows with 'Not assigned' in the 'Borough' column

In [4]:
print(df.shape)
df.head()

(180, 3)


Postcode           Borough               Neighborhood
0      M1A      Not assigned               Not assigned
1      M2A      Not assigned               Not assigned
2      M3A        North York                  Parkwoods
3      M4A        North York           Victoria Village
4      M5A  Downtown Toronto  Harbourfront, Regent Park

In [5]:
df_dropped = df[df['Borough'] != 'Not assigned']

In [6]:
df_dropped.reset_index(drop=True,inplace=True)

In [7]:
df_dropped.head()

Postcode           Borough                      Neighborhood
0      M3A        North York                         Parkwoods
1      M4A        North York                  Victoria Village
2      M5A  Downtown Toronto         Harbourfront, Regent Park
3      M6A        North York  Lawrence Heights, Lawrence Manor
4      M7A      Queen's Park                      Queen's Park

In [8]:
df_dropped.shape

(103, 3)

## Now for the method to use Pandas to read in the table from the url using pd.read_url()

In [9]:
weburl = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
new_df = pd.read_html(weburl)[0]  #The [0] at the end is necessary to pull the first part of the data read in which is the table

In [10]:
print(new_df.shape)
new_df.head()

(288, 3)


Postcode           Borough     Neighbourhood
0      M1A      Not assigned      Not assigned
1      M2A      Not assigned      Not assigned
2      M3A        North York         Parkwoods
3      M4A        North York  Victoria Village
4      M5A  Downtown Toronto      Harbourfront

In [11]:
new_df_drop = new_df[new_df['Borough'] != 'Not assigned']
new_df_drop.reset_index(drop=True,inplace=True)

In [12]:
def set_same_borough(x):
    for i in range(len(x['Neighbourhood'])):
        if x['Neighbourhood'][i]=='Not assigned':
            x['Neighbourhood'][i]=x['Borough'][i]
    return x

#I am going to set the same neighborhood as the borough if the neighborhood is not assigned
new_df_drop = set_same_borough(new_df_drop)
print(new_df_drop[new_df_drop['Neighbourhood']=='Not assigned'])

Empty DataFrame
Columns: [Postcode, Borough, Neighbourhood]
Index: []


C:\Users\KevinDWilk\Anaconda3\lib\site-packages\ipykernel_launcher.py:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  


In [13]:
new_df_drop.shape

(211, 3)

In [14]:
new_df_group = new_df_drop.groupby('Postcode')

In [15]:
def join_nghds(x):
    x['Neighbourhood'] = ', '.join(x['Neighbourhood'])
    return x
final_df=new_df_group.apply(join_nghds)

In [16]:
final_df.head()

Postcode           Borough                     Neighbourhood
0      M3A        North York                         Parkwoods
1      M4A        North York                  Victoria Village
2      M5A  Downtown Toronto         Harbourfront, Regent Park
3      M5A  Downtown Toronto         Harbourfront, Regent Park
4      M6A        North York  Lawrence Heights, Lawrence Manor

In [17]:
final_df.drop_duplicates(subset='Postcode',keep='first',inplace=True)
final_df.head()

Postcode           Borough                     Neighbourhood
0      M3A        North York                         Parkwoods
1      M4A        North York                  Victoria Village
2      M5A  Downtown Toronto         Harbourfront, Regent Park
4      M6A        North York  Lawrence Heights, Lawrence Manor
6      M7A      Queen's Park                      Queen's Park

In [18]:
print('Note it is the same as the previous method...YAY!')
final_df.shape

Note it is the same as the previous method...YAY!


(103, 3)

# Now on to the second part

## Remember the goal of this part is to read in the coordinates information and make the table.

Since this method doesn't seem to be working, I am going to just load in the datasets and merge them.

In [19]:
coords_df = pd.read_csv('Geospatial_Coordinates.csv')

In [20]:
coords_df.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [21]:
coords_df.shape

(103, 3)

They have the same length (103 rows), so I will assume that it has all the same Postal Codes that I have

In [22]:
coords_df.rename(columns={'Postal Code':'Postcode'},inplace=True)

In [23]:
joined_df = pd.merge(final_df,coords_df, on='Postcode',how='inner')

In [24]:
joined_df

Postcode           Borough  \
0        M3A        North York   
1        M4A        North York   
2        M5A  Downtown Toronto   
3        M6A        North York   
4        M7A      Queen's Park   
5        M9A         Etobicoke   
6        M1B       Scarborough   
7        M3B        North York   
8        M4B         East York   
9        M5B  Downtown Toronto   
10       M6B        North York   
11       M9B         Etobicoke   
12       M1C       Scarborough   
13       M3C        North York   
14       M4C         East York   
15       M5C  Downtown Toronto   
16       M6C              York   
17       M9C         Etobicoke   
18       M1E       Scarborough   
19       M4E      East Toronto   
20       M5E  Downtown Toronto   
21       M6E              York   
22       M1G       Scarborough   
23       M4G         East York   
24       M5G  Downtown Toronto   
25       M6G  Downtown Toronto   
26       M1H       Scarborough   
27       M2H        North York   
28       M3H        North York   
29       M4H         East York   
30       M5H  Downtown Toronto   
31       M6H      West Toronto   
32       M1J       Scarborough   
33       M2J        North York   
34       M3J        North York   
35       M4J         East York   
36       M5J  Downtown Toronto   
37       M6J      West Toronto   
38       M1K       Scarborough   
39       M2K        North York   
40       M3K        North York   
41       M4K      East Toronto   
42       M5K  Downtown Toronto   
43       M6K      West Toronto   
44       M1L       Scarborough   
45       M2L        North York   
46       M3L        North York   
47       M4L      East Toronto   
48       M5L  Downtown Toronto   
49       M6L        North York   
50       M9L        North York   
51       M1M       Scarborough   
52       M2M        North York   
53       M3M        North York   
54       M4M      East Toronto   
55       M5M        North York   
56       M6M              York   
57       M9M        North York   
58       M1N       Scarborough   
59       M2N        North York   
60       M3N        North York   
61       M4N   Central Toronto   
62       M5N   Central Toronto   
63       M6N              York   
64       M9N              York   
65       M1P       Scarborough   
66       M2P        North York   
67       M4P   Central Toronto   
68       M5P   Central Toronto   
69       M6P      West Toronto   
70       M9P         Etobicoke   
71       M1R       Scarborough   
72       M2R        North York   
73       M4R   Central Toronto   
74       M5R   Central Toronto   
75       M6R      West Toronto   
76       M7R       Mississauga   
77       M9R         Etobicoke   
78       M1S       Scarborough   
79       M4S   Central Toronto   
80       M5S  Downtown Toronto   
81       M6S      West Toronto   
82       M1T       Scarborough   
83       M4T   Central Toronto   
84       M5T  Downtown Toronto   
85       M1V       Scarborough   
86       M4V   Central Toronto   
87       M5V  Downtown Toronto   
88       M8V         Etobicoke   
89       M9V         Etobicoke   
90       M1W       Scarborough   
91       M4W  Downtown Toronto   
92       M5W  Downtown Toronto   
93       M8W         Etobicoke   
94       M9W         Etobicoke   
95       M1X       Scarborough   
96       M4X  Downtown Toronto   
97       M5X  Downtown Toronto   
98       M8X         Etobicoke   
99       M4Y  Downtown Toronto   
100      M7Y      East Toronto   
101      M8Y         Etobicoke   
102      M8Z         Etobicoke   

                                         Neighbourhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656  
1                                     Victoria Village  43.725882 -79.315572  
2                            Harbourfront, Regent Park  43.654260 -79.360636  
3                     Lawrence Heights, Lawrence Manor  43.718518 -79.464763  
4                                         Queen's Park  43.662301 -79.38

We've now made the table for the coordinates for each postal code...

# Now for the third part.
## In this third part we apply clustering algorithms to the neighborhoods of Toronto, Canada

In [25]:
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto, Ontario are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto, Ontario are 43.653963, -79.387207.


In [26]:
# create map of New York using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(joined_df['Latitude'], joined_df['Longitude'], 
                                           joined_df['Borough'], joined_df['Neighbourhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

In [27]:
CLIENT_ID = 'UZNFARAB5RPK3VRKCEPIBJIU0YUBE4EMDMUOWL3FK2OBNLK4' # your Foursquare ID
CLIENT_SECRET = 'R3KLBEUMMU5A3FTOX4WW1SRJHKED1SQXGMSYTFXHFJ1PEXRE' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: UZNFARAB5RPK3VRKCEPIBJIU0YUBE4EMDMUOWL3FK2OBNLK4
CLIENT_SECRET:R3KLBEUMMU5A3FTOX4WW1SRJHKED1SQXGMSYTFXHFJ1PEXRE


### Since there are a lot of boroughs, I will restrict this to just boroughs with Toronto in the name.

In [28]:
toronto_data = joined_df[joined_df['Borough'].str.contains('Toronto')].reset_index(drop=True)

In [29]:
toronto_data.head()

Postcode           Borough              Neighbourhood   Latitude  Longitude
0      M5A  Downtown Toronto  Harbourfront, Regent Park  43.654260 -79.360636
1      M5B  Downtown Toronto   Ryerson, Garden District  43.657162 -79.378937
2      M5C  Downtown Toronto             St. James Town  43.651494 -79.375418
3      M4E      East Toronto                The Beaches  43.676357 -79.293031
4      M5E  Downtown Toronto                Berczy Park  43.644771 -79.373306

In [30]:
# create map of New York using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, borough, neighborhood in zip(toronto_data['Latitude'], toronto_data['Longitude'], 
                                           toronto_data['Borough'], toronto_data['Neighbourhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

In [31]:
toronto_data[toronto_data['Borough']=='Downtown Toronto']

Postcode           Borough  \
0       M5A  Downtown Toronto   
1       M5B  Downtown Toronto   
2       M5C  Downtown Toronto   
4       M5E  Downtown Toronto   
5       M5G  Downtown Toronto   
6       M6G  Downtown Toronto   
7       M5H  Downtown Toronto   
9       M5J  Downtown Toronto   
12      M5K  Downtown Toronto   
15      M5L  Downtown Toronto   
26      M5S  Downtown Toronto   
29      M5T  Downtown Toronto   
31      M5V  Downtown Toronto   
32      M4W  Downtown Toronto   
33      M5W  Downtown Toronto   
34      M4X  Downtown Toronto   
35      M5X  Downtown Toronto   
36      M4Y  Downtown Toronto   

                                        Neighbourhood   Latitude  Longitude  
0                           Harbourfront, Regent Park  43.654260 -79.360636  
1                            Ryerson, Garden District  43.657162 -79.378937  
2                                      St. James Town  43.651494 -79.375418  
4                                         Berczy Park  43.644771 -79.373306  
5                                  Central Bay Street  43.657952 -79.387383  
6                                            Christie  43.669542 -79.422564  
7                            Adelaide, King, Richmond  43.650571 -79.384568  
9   Harbourfront East, Toronto Islands, Union Station  43.640816 -79.381752  
12           Design Exchange, Toronto Dominion Centre  43.647177 -79.381576  
15                     Commerce Court, Victoria Hotel  43.648198 -79.379817  
26                     Harbord, University of Toronto  43.662696 -79.400049  
29          Chinatown, Grange Park, Kensington Market  43.653206 -79.400049  
31  CN Tower, Bathurst Quay, Island airport, Harbo...  43.628947 -79.394420  
32                                           Rosedale  43.679563 -79.377529  
33                    Stn A PO Boxes 25 The Esplanade  43.646435 -79.374846  
34                        Cabbagetown, St. James Town  43.667967 -79.367675  
35             First Canadian Place, Underground city  43.648429 -79.382280  
36                               Church and Wellesley  43.665860 -79.383160

In [32]:
toronto_data[toronto_data['Borough']=='Downtown Toronto'].reset_index(drop=True).loc[0,'Neighbourhood'].split(',')[0]

'Harbourfront'

In [33]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
        LIMIT = 100
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

### After using the function above for the call to the API, we can now call it using our Toronto data frame

In [34]:
toronto_venues = getNearbyVenues(names=toronto_data['Neighbourhood'],
                                   latitudes=toronto_data['Latitude'],
                                   longitudes=toronto_data['Longitude']
                                  )

Harbourfront, Regent Park
Ryerson, Garden District
St. James Town
The Beaches
Berczy Park
Central Bay Street
Christie
Adelaide, King, Richmond
Dovercourt Village, Dufferin
Harbourfront East, Toronto Islands, Union Station
Little Portugal, Trinity
The Danforth West, Riverdale
Design Exchange, Toronto Dominion Centre
Brockton, Exhibition Place, Parkdale Village
The Beaches West, India Bazaar
Commerce Court, Victoria Hotel
Studio District
Lawrence Park
Roselawn
Davisville North
Forest Hill North, Forest Hill West
High Park, The Junction South
North Toronto West
The Annex, North Midtown, Yorkville
Parkdale, Roncesvalles
Davisville
Harbord, University of Toronto
Runnymede, Swansea
Moore Park, Summerhill East
Chinatown, Grange Park, Kensington Market
Deer Park, Forest Hill SE, Rathnelly, South Hill, Summerhill West
CN Tower, Bathurst Quay, Island airport, Harbourfront West, King and Spadina, Railway Lands, South Niagara
Rosedale
Stn A PO Boxes 25 The Esplanade
Cabbagetown, St. James Town
Fir

In [35]:
print(toronto_venues.shape)
toronto_venues.head()

(1715, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  Harbourfront, Regent Park               43.65426              -79.360636   
1  Harbourfront, Regent Park               43.65426              -79.360636   
2  Harbourfront, Regent Park               43.65426              -79.360636   
3  Harbourfront, Regent Park               43.65426              -79.360636   
4  Harbourfront, Regent Park               43.65426              -79.360636   

                                             Venue  Venue Latitude  \
0                                 Roselle Desserts       43.653447   
1                                    Tandem Coffee       43.653559   
2  Toronto Cooper Koo Family Cherry St YMCA Centre       43.653191   
3                              Body Blitz Spa East       43.654735   
4                               Morning Glory Cafe       43.653947   

   Venue Longitude        Venue Category  
0       -79.362017                Bakery  
1       -79.361809           Coffee Shop  
2       -79.357947  Gym / Fitness Center  
3       -79.359874                   Spa  
4       -79.361149        Breakfast Spot

In [36]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Adelaide, King, Richmond                                              100   
Berczy Park                                                            57   
Brockton, Exhibition Place, Parkdale Village                           23   
Business Reply Mail Processing Centre 969 Eastern                      19   
CN Tower, Bathurst Quay, Island airport, Harbou...                     14   
Cabbagetown, St. James Town                                            45   
Central Bay Street                                                     89   
Chinatown, Grange Park, Kensington Market                             100   
Christie                                                               15   
Church and Wellesley                                                   85   
Commerce Court, Victoria Hotel                                        100   
Davisville                                                             35   
Davisville North                                                        7   
Deer Park, Forest Hill SE, Rathnelly, South Hil...                     16   
Design Exchange, Toronto Dominion Centre                              100   
Dovercourt Village, Dufferin                                           15   
First Canadian Place, Underground city                                100   
Forest Hill North, Forest Hill West                                     4   
Harbord, University of Toronto                                         37   
Harbourfront East, Toronto Islands, Union Station                     100   
Harbourfront, Regent Park                                              53   
High Park, The Junction South                                          23   
Lawrence Park                                                           3   
Little Portugal, Trinity                                               65   
Moore Park, Summerhill East                                             4   
North Toronto West                                                     22   
Parkdale, Roncesvalles                                                 15   
Rosedale                                                                5   
Roselawn                                                                3   
Runnymede, Swansea                                                     35   
Ryerson, Garden District                                              100   
St. James Town                                                        100   
Stn A PO Boxes 25 The Esplanade                                        99   
Studio District                                                        40   
The Annex, North Midtown, Yorkville                                    21   
The Beaches                                                             4   
The Beaches West, India Bazaar                                         20   
The Danforth West, Riverdale                                           42   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Adelaide, King, Richmond                                               100   
Berczy Park                                                             57   
Brockton, Exhibition Place, Parkdale Village                            23   
Business Reply Mail Processing Centre 969 Eastern                       19   
CN Tower, Bathurst Quay, Island airport, Harbou...                      14   
Cabbagetown, St. James Town                                             45   
Central Bay Street                                                      89   
Chinatown, Grange Park, Kensington Market                              100   
Christie                                                                15   
Church and Wellesley                                                    85   
Commerce Court, Victoria Hotel     

In [37]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 238 uniques categories.


In [38]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Yoga Studio  Afghan Restaurant  Airport  Airport Food Court  \
0            0                  0        0                   0   
1            0                  0        0                   0   
2            0                  0        0                   0   
3            0                  0        0                   0   
4            0                  0        0                   0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  Aquarium  Art Gallery  Art Museum  Arts & Crafts Store  \
0             0         0            0           0                    0   
1             0         0            0           0                    0   
2             0         0            0           0                    0   
3             0         0            0           0                    0   
4             0         0            0           0                    0   

   Asian Restaurant  Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  Bakery  \
0                 0              0          0           0           0       1   
1                 0              0          0           0           0       0   
2                 0              0          0           0           0       0   
3                 0              0          0           0           0       0   
4                 0              0          0           0           0       0   

   Bank  Bar  Baseball Stadium  Basketball Stadium  Beach  Bed & Breakfast  \
0     0    0                 0                   0      0                0   
1     0    0                 0                   0      0                0   
2     0    0                 0                   0      0                0   
3     0    0                 0                   0      0                0   
4     0    0                 0                   0      0                0   

   Beer Bar  Beer Store  Belgian Restaurant  Bistro  Board Shop  \
0         0           0                   0       0           0   
1         0           0                   0       0           0   
2         0           0                   0       0           0   
3         0           0                   0       0           0   
4         0           0                   0       0           0   

   Boat or Ferry  Bookstore  Boutique  Brazilian Restaurant  Breakfast Spot  \
0              0          0         0                     0               0   
1              0          0         0                     0               0   
2              0          0         0                     0               0   
3              0          0         0                     0               0   
4              0          0         0                     0               1   

   Brewery  Bubble Tea Shop  Building  Burger Joint  Burrito Place  Bus Line  \
0        0                0         0             0              0         0   
1        0                0         0             0              0         0   
2        0                0         0             0              0         0   
3        0                0         0             0              0         0   
4        0                0         0             0              0         0   

   Butcher  Café  Cajun / Creole Restaurant  Camera Store  Candy Store  \
0        0     0                          0             0            0   
1        0     0                          0             0            0   
2        0     0                          0             0            0   
3        0     0                          0             0            0   
4      

In [39]:
toronto_onehot.shape

(1715, 238)

In [40]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  Yoga Studio  \
0                            Adelaide, King, Richmond     0.000000   
1                                         Berczy Park     0.000000   
2        Brockton, Exhibition Place, Parkdale Village     0.000000   
3   Business Reply Mail Processing Centre 969 Eastern     0.052632   
4   CN Tower, Bathurst Quay, Island airport, Harbo...     0.000000   
5                         Cabbagetown, St. James Town     0.000000   
6                                  Central Bay Street     0.011236   
7           Chinatown, Grange Park, Kensington Market     0.000000   
8                                            Christie     0.000000   
9                                Church and Wellesley     0.023529   
10                     Commerce Court, Victoria Hotel     0.000000   
11                                         Davisville     0.000000   
12                                   Davisville North     0.000000   
13  Deer Park, Forest Hill SE, Rathnelly, South Hi...     0.000000   
14           Design Exchange, Toronto Dominion Centre     0.000000   
15                       Dovercourt Village, Dufferin     0.000000   
16             First Canadian Place, Underground city     0.000000   
17                Forest Hill North, Forest Hill West     0.000000   
18                     Harbord, University of Toronto     0.027027   
19  Harbourfront East, Toronto Islands, Union Station     0.000000   
20                          Harbourfront, Regent Park     0.018868   
21                      High Park, The Junction South     0.000000   
22                                      Lawrence Park     0.000000   
23                           Little Portugal, Trinity     0.015385   
24                        Moore Park, Summerhill East     0.000000   
25                                 North Toronto West     0.045455   
26                             Parkdale, Roncesvalles     0.000000   
27                                           Rosedale     0.000000   
28                                           Roselawn     0.000000   
29                                 Runnymede, Swansea     0.000000   
30                           Ryerson, Garden District     0.000000   
31                                     St. James Town     0.000000   
32                    Stn A PO Boxes 25 The Esplanade     0.000000   
33                                    Studio District     0.050000   
34                The Annex, North Midtown, Yorkville     0.000000   
35                                        The Beaches     0.000000   
36                     The Beaches West, India Bazaar     0.000000   
37                       The Danforth West, Riverdale     0.023810   

    Afghan Restaurant   Airport  Airport Food Court  Airport Lounge  \
0            0.000000  0.000000            0.000000        0.000000   
1            0.000000  0.000000            0.000000        0.000000   
2            0.000000  0.000000            0.000000        0.000000   
3            0.000000  0.000000            0.000000        0.000000   
4            0.000000  0.071429            0.071429        0.142857   
5            0.000000  0.000000            0.000000        0.000000   
6            0.000000  0.000000            0.000000        0.000000   
7            0.000000  0.000000            0.000000        0.000000   
8            0.000000  0.000000            0.000000        0.000000   
9            0.011765  0.000000            0.000000        0.000000   
10           0.000000  0.000000            0.000000        0.000000   
11           0.000000  0.000000            0.000000        0.000000   
12           0.000000  0.000000            0.000000        0.000000   
13           0.000000  0.000000            0.000000        0.000000   
14           0.000000  0.000000            0.000000        0.000000   
15           0.000000  0.000000            0.000000        0.000000   
16           0.000000  0.000000            0.000000        0.000000   
17           0.000000  0.000000 

In [41]:
toronto_grouped.shape

(38, 238)

In [42]:
num_top_venues = 5

for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Adelaide, King, Richmond----
             venue  freq
0      Coffee Shop  0.07
1             Café  0.05
2       Steakhouse  0.04
3              Bar  0.04
4  Thai Restaurant  0.04


----Berczy Park----
                venue  freq
0         Coffee Shop  0.07
1        Cocktail Bar  0.05
2          Steakhouse  0.04
3      Farmers Market  0.04
4  Seafood Restaurant  0.04


----Brockton, Exhibition Place, Parkdale Village----
               venue  freq
0               Café  0.13
1     Breakfast Spot  0.09
2        Coffee Shop  0.09
3  Convenience Store  0.04
4         Restaurant  0.04


----Business Reply Mail Processing Centre 969 Eastern----
                venue  freq
0  Light Rail Station  0.11
1         Yoga Studio  0.05
2                Park  0.05
3          Comic Shop  0.05
4    Recording Studio  0.05


----CN Tower, Bathurst Quay, Island airport, Harbourfront West, King and Spadina, Railway Lands, South Niagara----
              venue  freq
0    Airport Lounge  0.14
1  Airport Te

In [43]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [44]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                           Adelaide, King, Richmond           Coffee Shop   
1                                        Berczy Park           Coffee Shop   
2       Brockton, Exhibition Place, Parkdale Village                  Café   
3  Business Reply Mail Processing Centre 969 Eastern    Light Rail Station   
4  CN Tower, Bathurst Quay, Island airport, Harbo...        Airport Lounge   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                  Café                   Bar            Steakhouse   
1          Cocktail Bar           Cheese Shop                  Café   
2        Breakfast Spot           Coffee Shop                   Gym   
3           Yoga Studio         Auto Workshop                  Park   
4      Airport Terminal                 Plane                   Bar   

    5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0         Thai Restaurant            Restaurant          Burger Joint   
1              Steakhouse                Bakery        Farmers Market   
2  Furniture / Home Store         Burrito Place     Convenience Store   
3             Pizza Place      Recording Studio            Restaurant   
4             Coffee Shop      Sculpture Garden              Boutique   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0      Sushi Restaurant      Asian Restaurant    American Restaurant  
1              Beer Bar    Seafood Restaurant          Shopping Mall  
2        Sandwich Place               Stadium             Restaurant  
3               Butcher         Burrito Place                Brewery  
4         Boat or Ferry       Harbor / Marina                Airport

## Cluster Neighborhoods

In [45]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=1).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:20] 

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 4, 0, 0])

In [46]:
toronto_grouped_clustering.head()

Yoga Studio  Afghan Restaurant   Airport  Airport Food Court  \
0     0.000000                0.0  0.000000            0.000000   
1     0.000000                0.0  0.000000            0.000000   
2     0.000000                0.0  0.000000            0.000000   
3     0.052632                0.0  0.000000            0.000000   
4     0.000000                0.0  0.071429            0.071429   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0        0.000000         0.000000          0.000000                 0.03   
1        0.000000         0.000000          0.000000                 0.00   
2        0.000000         0.000000          0.000000                 0.00   
3        0.000000         0.000000          0.000000                 0.00   
4        0.142857         0.071429          0.142857                 0.00   

   Antique Shop  Aquarium  Art Gallery  Art Museum  Arts & Crafts Store  \
0           0.0       0.0     0.010000        0.01                  0.0   
1           0.0       0.0     0.017544        0.00                  0.0   
2           0.0       0.0     0.000000        0.00                  0.0   
3           0.0       0.0     0.000000        0.00                  0.0   
4           0.0       0.0     0.000000        0.00                  0.0   

   Asian Restaurant  Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  \
0              0.03       0.000000   0.000000         0.0    0.000000   
1              0.00       0.000000   0.017544         0.0    0.017544   
2              0.00       0.000000   0.000000         0.0    0.000000   
3              0.00       0.052632   0.000000         0.0    0.000000   
4              0.00       0.000000   0.000000         0.0    0.000000   

     Bakery  Bank       Bar  Baseball Stadium  Basketball Stadium     Beach  \
0  0.020000   0.0  0.040000               0.0            0.000000  0.000000   
1  0.035088   0.0  0.000000               0.0            0.017544  0.017544   
2  0.043478   0.0  0.043478               0.0            0.000000  0.000000   
3  0.000000   0.0  0.000000               0.0            0.000000  0.000000   
4  0.000000   0.0  0.071429               0.0            0.000000  0.000000   

   Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant    Bistro  \
0              0.0  0.000000         0.0            0.000000  0.000000   
1              0.0  0.035088         0.0            0.017544  0.017544   
2              0.0  0.000000         0.0            0.000000  0.000000   
3              0.0  0.000000         0.0            0.000000  0.000000   
4              0.0  0.000000         0.0            0.000000  0.000000   

   Board Shop  Boat or Ferry  Bookstore  Boutique  Brazilian Restaurant  \
0         0.0       0.000000       0.01  0.000000                  0.01   
1         0.0       0.000000       0.00  0.000000                  0.00   
2         0.0       0.000000       0.00  0.000000                  0.00   
3         0.0       0.000000       0.00  0.000000                  0.00   
4         0.0       0.071429       0.00  0.071429                  0.00   

   Breakfast Spot   Brewery  Bubble Tea Shop  Building  Burger Joint  \
0        0.020000  0.000000              0.0      0.01          0.03   
1        0.017544  0.000000              0.0      0.00          0.00   
2        0.086957  0.000000              0.0      0.00          0.00   
3        0.000000  0.052632              0.0      0.00          0.00   
4        0.000000  0.000000              0.0      0.00          0.00   

   Burrito Place  Bus Line   Butcher      Café  Cajun / Creole Restaurant  \
0       0.010000       0.0  0.000000  0.050000                        0.0   
1       0.000000       0.0  0.000000  0.035088                        0.0   
2       0.043478       0.0  0.000000  0.130435                        0.0   
3       0.052632       0.0  0.052632  0.000000                        0.0   
4       0.000000       0.0  0.000000  0.000000            

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [47]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

In [50]:
toronto_merged = toronto_data
toronto_merged.head()

Postcode           Borough              Neighbourhood   Latitude  Longitude
0      M5A  Downtown Toronto  Harbourfront, Regent Park  43.654260 -79.360636
1      M5B  Downtown Toronto   Ryerson, Garden District  43.657162 -79.378937
2      M5C  Downtown Toronto             St. James Town  43.651494 -79.375418
3      M4E      East Toronto                The Beaches  43.676357 -79.293031
4      M5E  Downtown Toronto                Berczy Park  43.644771 -79.373306

In [51]:
# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighbourhood')

toronto_merged.head() # check the last columns!

Postcode           Borough              Neighbourhood   Latitude  Longitude  \
0      M5A  Downtown Toronto  Harbourfront, Regent Park  43.654260 -79.360636   
1      M5B  Downtown Toronto   Ryerson, Garden District  43.657162 -79.378937   
2      M5C  Downtown Toronto             St. James Town  43.651494 -79.375418   
3      M4E      East Toronto                The Beaches  43.676357 -79.293031   
4      M5E  Downtown Toronto                Berczy Park  43.644771 -79.373306   

   Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0               0           Coffee Shop                  Park   
1               0           Coffee Shop        Clothing Store   
2               0           Coffee Shop                  Café   
3               0     Health Food Store                 Trail   
4               0           Coffee Shop          Cocktail Bar   

  3rd Most Common Venue      4th Most Common Venue 5th Most Common Venue  \
0                   Pub                     Bakery                  Café   
1        Cosmetics Shop  Middle Eastern Restaurant    Italian Restaurant   
2            Restaurant                      Hotel        Clothing Store   
3                   Pub         Dim Sum Restaurant        Farmers Market   
4           Cheese Shop                       Café            Steakhouse   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0    Mexican Restaurant        Breakfast Spot               Theater   
1                  Café        Lingerie Store       Bubble Tea Shop   
2             Gastropub        Breakfast Spot    Italian Restaurant   
3    Falafel Restaurant           Event Space  Ethiopian Restaurant   
4                Bakery        Farmers Market              Beer Bar   

  9th Most Common Venue       10th Most Common Venue  
0         Historic Site         Gym / Fitness Center  
1      Ramen Restaurant          Sporting Goods Shop  
2                Bakery               Cosmetics Shop  
3     Electronics Store  Eastern European Restaurant  
4    Seafood Restaurant                Shopping Mall

In [52]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], 
                                  toronto_merged['Neighbourhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

### We can now look at our clusters

### Cluster 1

In [53]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0   Downtown Toronto               0           Coffee Shop   
1   Downtown Toronto               0           Coffee Shop   
2   Downtown Toronto               0           Coffee Shop   
3       East Toronto               0     Health Food Store   
4   Downtown Toronto               0           Coffee Shop   
5   Downtown Toronto               0           Coffee Shop   
6   Downtown Toronto               0         Grocery Store   
7   Downtown Toronto               0           Coffee Shop   
8       West Toronto               0                Bakery   
9   Downtown Toronto               0           Coffee Shop   
10      West Toronto               0                   Bar   
11      East Toronto               0      Greek Restaurant   
12  Downtown Toronto               0           Coffee Shop   
13      West Toronto               0                  Café   
14      East Toronto               0                  Park   
15  Downtown Toronto               0           Coffee Shop   
16      East Toronto               0                  Café   
19   Central Toronto               0                 Hotel   
21      West Toronto               0                  Café   
22   Central Toronto               0        Clothing Store   
23   Central Toronto               0                  Café   
24      West Toronto               0        Breakfast Spot   
25   Central Toronto               0          Dessert Shop   
26  Downtown Toronto               0                  Café   
27      West Toronto               0           Coffee Shop   
29  Downtown Toronto               0                  Café   
30   Central Toronto               0                   Pub   
31  Downtown Toronto               0        Airport Lounge   
33  Downtown Toronto               0           Coffee Shop   
34  Downtown Toronto               0           Coffee Shop   
35  Downtown Toronto               0           Coffee Shop   
36  Downtown Toronto               0           Coffee Shop   
37      East Toronto               0    Light Rail Station   

            2nd Most Common Venue 3rd Most Common Venue  \
0                            Park                   Pub   
1                  Clothing Store        Cosmetics Shop   
2                            Café            Restaurant   
3                           Trail                   Pub   
4                    Cocktail Bar           Cheese Shop   
5              Italian Restaurant                  Café   
6                            Café                  Park   
7                            Café                   Bar   
8                        Pharmacy           Supermarket   
9                           Hotel              Aquarium   
10                    Coffee Shop           Men's Store   
11                    Coffee Shop        Ice Cream Shop   
12                           Café                 Hotel   
13                 Breakfast Spot           Coffee Shop   
14                      Pet Store           Coffee Shop   
15                           Café                 Hotel   
16                    Coffee Shop           Yoga Studio   
19              Food & Drink Shop                  Park   
21             Mexican Restaurant       Thai Restaurant   
22                    Coffee Shop   Sporting Goods Shop   
23                 Sandwich Place           Coffee Shop   
24                      Gift Shop          Dessert Shop   
25                 Sandwich Place                   Gym   
26                    Coffee Shop                Bakery   
27                           Café           Pizza Place   
29  Vegetarian / Vegan Restaurant                   Bar   
30                    Coffee Shop            Bagel Shop   
31               Airport Terminal                 Plane   
33                     Restaurant                  Café   
34                           Park            Restaurant   
35                           Café            Steakhouse   
36              

### Cluster 2

In [54]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
17  Central Toronto               1                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
17           Swim School              Bus Line         Women's Store   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
17                 Diner    Falafel Restaurant           Event Space   

   8th Most Common Venue 9th Most Common Venue       10th Most Common Venue  
17  Ethiopian Restaurant     Electronics Store  Eastern European Restaurant

### Cluster 3

In [55]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
28  Central Toronto               2                   Gym   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
28            Playground                 Trail          Tennis Court   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
28         Women's Store            Donut Shop                 Diner   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
28        Discount Store               Dog Run       Doner Restaurant

### Cluster 4

In [56]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
18  Central Toronto               3                Garden   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
18          Home Service        Ice Cream Shop         Women's Store   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
18        Farmers Market    Falafel Restaurant           Event Space   

   8th Most Common Venue 9th Most Common Venue       10th Most Common Venue  
18  Ethiopian Restaurant     Electronics Store  Eastern European Restaurant

### Cluster 5

In [57]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
20   Central Toronto               4                  Park   
32  Downtown Toronto               4                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
20      Sushi Restaurant         Jewelry Store                 Trail   
32            Playground                 Trail              Building   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
20         Women's Store            Donut Shop        Discount Store   
32         Women's Store   Dumpling Restaurant        Discount Store   

   8th Most Common Venue 9th Most Common Venue       10th Most Common Venue  
20               Dog Run      Doner Restaurant  Eastern European Restaurant  
32               Dog Run      Doner Restaurant                   Donut Shop